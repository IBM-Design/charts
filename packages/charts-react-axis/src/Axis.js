import React, { PropTypes } from 'react';
import Tick from './Tick';

/* eslint-disable react/display-name */
const calcLabels = (max, min, tickCount) => {
  return Array.from(Array(tickCount + 1)).map((_, i) => {
    return i * (max - min) / tickCount + min;
  });
};

const renderTickWith = (height, interval, labels, length, type) => (_, i) => {
  const props = type === 'bottom'
    ? {
      dy: 0.71,
      offset: `translate(${interval * i}, 0)`,
      path: `M0,0 L0,-${length}`,
      x: 0.5,
      y: 9,
    }
    : {
      dy: 0.32,
      offset: `translate(0, ${height - (interval * i)})`,
      path: `M0,0 L${length},0`,
      x: -9,
      y: 0.5,
    };

  return (
    <Tick
      key={i}
      label={labels[i].toString()}
      {...props}
    />
  );
};
/* eslint-enable react/display-name */

const Axis = (props, context) => {
  const {
    max,
    min,
    text,
    tickCount,
    type,
  } = props;
  const { chart } = context;
  const length = type === 'bottom'
    ? chart.height
    : chart.width;
  const totalTicks = text
    ? text.length - 1
    : tickCount;
  const interval =  type === 'bottom'
    ? chart.width / totalTicks
    : chart.height / totalTicks;
  const labels = text
    ? text
    : calcLabels(max, min, tickCount);
  const transform = type === 'bottom'
    ? `translate(0,${chart.height})`
    : '';

  const renderTick = renderTickWith(chart.height, interval, labels, length, type);

  return (
    <g
      className="axis axis--x"
      fill="none"
      fontSize="10"
      fontFamily="sans-serif"
      textAnchor="middle"
      transform={transform}>
      {Array.from(Array(totalTicks + 1)).map(renderTick)}
    </g>
  );
};

Axis.propTypes = {
  max: PropTypes.number,
  min: PropTypes.number,
  text: PropTypes.array, // eslint-disable-line react/forbid-prop-types
  tickCount: PropTypes.number,
  type: PropTypes.string.isRequired,
};

Axis.contextTypes = {
  chart: PropTypes.object,
};

Axis.defaultProps = {
  min: 0,
  type: 'bottom',
};

export default Axis;
