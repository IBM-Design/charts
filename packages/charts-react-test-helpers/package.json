{
  "name": "@ibm-design/charts-react-test-helpers",
  "version": "0.0.0",
  "main": "lib/index.js",
  "scripts": {
    "build": "npm run clean && babel -d lib src",
    "clean": "rimraf lib"
  },
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": "^15.4.2"
  },
  "devDependencies": {
    "babel-cli": "^6.18.0",
    "react": "^15.4.2",
    "rimraf": "^2.5.4"
  },
  "dependencies": {
    "prop-types": "^15.6.0"
  }
}
