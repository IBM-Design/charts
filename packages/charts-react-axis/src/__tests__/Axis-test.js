import React from 'react';
import renderer from 'react-test-renderer';
import { ChartProvider } from '@ibm-design/charts-react-test-helpers';
import Axis from '../Axis';

describe('Axis Component', () => {
  it('should render', () => {
    const tree = renderer.create(
      <ChartProvider>
        <Axis tickCount={5} />
      </ChartProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
