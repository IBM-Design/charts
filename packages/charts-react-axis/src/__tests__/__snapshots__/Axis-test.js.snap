exports[`Axis Component should render 1`] = `
<g
  className="axis axis--x"
  fill="none"
  fontFamily="sans-serif"
  fontSize="10"
  textAnchor="middle"
  transform="translate(0,500)">
  <g
    className="tick"
    transform="translate(0, 0)">
    <path
      d="M0,0 L0,-500"
      fill="none"
      stroke="#CDDBF1"
      strokeWidth="1"
      style={
        Object {
          "animation": "dashoffset 0.4s linear 0.3s forwards",
          "strokeDasharray": 500,
          "strokeDashoffset": 500,
        }
      } />
    <text
      dy="0.71em"
      fill="#000"
      x={0.5}
      y={9}>
      NaN
    </text>
  </g>
  <g
    className="tick"
    transform="translate(100, 0)">
    <path
      d="M0,0 L0,-500"
      fill="none"
      stroke="#CDDBF1"
      strokeWidth="1"
      style={
        Object {
          "animation": "dashoffset 0.4s linear 0.35s forwards",
          "strokeDasharray": 500,
          "strokeDashoffset": 500,
        }
      } />
    <text
      dy="0.71em"
      fill="#000"
      x={0.5}
      y={9}>
      NaN
    </text>
  </g>
  <g
    className="tick"
    transform="translate(200, 0)">
    <path
      d="M0,0 L0,-500"
      fill="none"
      stroke="#CDDBF1"
      strokeWidth="1"
      style={
        Object {
          "animation": "dashoffset 0.4s linear 0.4s forwards",
          "strokeDasharray": 500,
          "strokeDashoffset": 500,
        }
      } />
    <text
      dy="0.71em"
      fill="#000"
      x={0.5}
      y={9}>
      NaN
    </text>
  </g>
  <g
    className="tick"
    transform="translate(300, 0)">
    <path
      d="M0,0 L0,-500"
      fill="none"
      stroke="#CDDBF1"
      strokeWidth="1"
      style={
        Object {
          "animation": "dashoffset 0.4s linear 0.44999999999999996s forwards",
          "strokeDasharray": 500,
          "strokeDashoffset": 500,
        }
      } />
    <text
      dy="0.71em"
      fill="#000"
      x={0.5}
      y={9}>
      NaN
    </text>
  </g>
  <g
    className="tick"
    transform="translate(400, 0)">
    <path
      d="M0,0 L0,-500"
      fill="none"
      stroke="#CDDBF1"
      strokeWidth="1"
      style={
        Object {
          "animation": "dashoffset 0.4s linear 0.5s forwards",
          "strokeDasharray": 500,
          "strokeDashoffset": 500,
        }
      } />
    <text
      dy="0.71em"
      fill="#000"
      x={0.5}
      y={9}>
      NaN
    </text>
  </g>
  <g
    className="tick"
    transform="translate(500, 0)">
    <path
      d="M0,0 L0,-500"
      fill="none"
      stroke="#CDDBF1"
      strokeWidth="1"
      style={
        Object {
          "animation": "dashoffset 0.4s linear 0.55s forwards",
          "strokeDasharray": 500,
          "strokeDashoffset": 500,
        }
      } />
    <text
      dy="0.71em"
      fill="#000"
      x={0.5}
      y={9}>
      NaN
    </text>
  </g>
</g>
`;
